ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,885 @@
+  private final DiskIOScheduler diskIOScheduler;
+  private final ScheduledExecutorService scheduler;
+  private final MessageStoreRecovery recovery;
+  private final Time time;
+  private final UUID sessionId;
+  private final UUID incarnationId;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  private volatile boolean isActive = false;
+  private PersistentIndex srcIndex;
+
+  private Log tgtLog;
+  private PersistentIndex tgtIndex;
+  private long numSwapsUsed;
+  private StoreFindToken recoveryStartToken = null;
+  private CompactionLog compactionLog;
+  private volatile CountDownLatch runningLatch = new CountDownLatch(0);
+
+  /**
+   * Constructs the compactor component.
+   * @param dataDir the directory where all compactions need to be run.
+   * @param storeId the unique ID of the store.
+   * @param storeKeyFactory the {@link StoreKeyFactory} to generate {@link StoreKey} instances.
+   * @param config the {@link StoreConfig} that defines configuration parameters.
+   * @param metrics the {@link StoreMetrics} to use to record metrics.
+   * @param diskIOScheduler the {@link DiskIOScheduler} to schedule I/O.
+   * @param srcLog the {@link Log} to copy data from.
+   * @param scheduler the {@link ScheduledExecutorService} to use.
+   * @param recovery the {@link MessageStoreRecovery} to use to recover the index and log.
+   * @param time the {@link Time} instance to use.
+   * @param sessionId the sessionID of the store.
+   * @param incarnationId the incarnation ID of the store.
+   * @throws IOException if the {@link CompactionLog} could not be created or if commit/cleanup failed during recovery.
+   * @throws StoreException if the commit failed during recovery.
+   */
+  BlobStoreCompactor(String dataDir, String storeId, StoreKeyFactory storeKeyFactory, StoreConfig config,
+      StoreMetrics metrics, DiskIOScheduler diskIOScheduler, Log srcLog, ScheduledExecutorService scheduler,
+      MessageStoreRecovery recovery, Time time, UUID sessionId, UUID incarnationId) throws IOException, StoreException {
+    this.dataDir = new File(dataDir);
+    this.storeId = storeId;
+    this.storeKeyFactory = storeKeyFactory;
+    this.config = config;
+    this.srcMetrics = metrics;
+    tgtMetrics = new StoreMetrics(storeId + METRICS_SUFFIX, metrics.getRegistry());
+    this.srcLog = srcLog;
+    this.diskIOScheduler = diskIOScheduler;
+    this.scheduler = scheduler;
+    this.recovery = recovery;
+    this.time = time;
+    this.sessionId = sessionId;
+    this.incarnationId = incarnationId;
+    fixStateIfRequired();
+  }
+
+  /**
+   * Initializes the compactor and sets the {@link PersistentIndex} to copy data from.
+   * @param srcIndex the {@link PersistentIndex} to copy data from.
+   */
+  void initialize(PersistentIndex srcIndex) {
+    this.srcIndex = srcIndex;
+    if (compactionLog == null && srcIndex.hardDeleter != null && srcIndex.hardDeleter.isPaused()) {
+      srcIndex.hardDeleter.resume();
```

Comment:
This is in place for a very specific scenario (which is tested in `interruptionDuringLastCommitTest()`).
1. Server crashes when compaction is in the commit/cleanup stage of the very last cycle of compaction (hard delete is still paused).
2. On construction, `fixStateIfRequired()` completes the compaction but cannot resume hard deletes because `srcIndex` is not available (because it has not been created).
3. On initialization, there is no compaction in progress (because `fixStateIfRequired()` completed the compaction), so hard delete needs to be restarted.